src/crimson/common/errorator.h

```diff
@@ -79,7 +79,7 @@ class error_t {
 template <class ErrorT, ErrorT ErrorV>
 struct unthrowable_wrapper : error_t<unthrowable_wrapper<ErrorT, ErrorV>> {
   unthrowable_wrapper(const unthrowable_wrapper&) = delete;
-  static constexpr unthrowable_wrapper instance{};
+  static const unthrowable_wrapper instance;
```

Comment:
i think the existing code complies to C++17. as `constexpr` implies inline and can be defined in-class.

Reply:
The problem is related to the "unthrowable_wrapper" not yet being fully constructed at that point.
CLANG is correct here, and gcc shouldn't have allowed it.

Reply:
could you share the error message and the version of clang you are using?

Reply:
** Below. Also take a look at https://www.godbolt.org/z/-UGAbb **

../src/crimson/common/errorator.h:82:47: error: constexpr variable cannot have non-literal type 'const crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec>'
  static inline constexpr unthrowable_wrapper instance{};
                                              ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/9/../../../../include/c++/9/type_traits:1302:66: note: in instantiation of template class 'crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec>' requested here
    : public integral_constant<bool, __is_base_of(_Base, _Derived)>
                                                                 ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/9/../../../../include/c++/9/type_traits:2962:40: note: in instantiation of template class 'std::is_base_of<crimson::error_t<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >, crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >' requested here
  inline constexpr bool is_base_of_v = is_base_of<_Base, _Derived>::value;
                                       ^
../src/crimson/common/errorator.h:274:50: note: in instantiation of variable template specialization 'std::is_base_of_v<crimson::error_t<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >, crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >' requested here
  static inline constexpr bool is_error_v = std::is_base_of_v<error_t<T>, T>;
                                                 ^
../src/crimson/common/errorator.h:276:25: note: in instantiation of static data member 'crimson::errorator<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >::is_error_v<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >' requested here
  static_assert((... && is_error_v<AllowedErrors>),
                        ^
../src/crimson/os/futurized_store.h:46:11: note: in instantiation of template class 'crimson::errorator<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >' requested here
  virtual read_errorator::future<ceph::bufferlist> read(
          ^
../src/crimson/common/errorator.h:82:47: note: incomplete type 'const crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec>' is not a literal type
  static inline constexpr unthrowable_wrapper instance{};
                                              ^
../src/crimson/common/errorator.h:80:8: note: definition of 'crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec>' is not complete until the closing '}'
struct unthrowable_wrapper : error_t<unthrowable_wrapper<ErrorT, ErrorV>> {
       ^
../src/crimson/common/errorator.h:276:3: error: static_assert failed due to requirement 'is_error_v<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >' "errorator expects presence of ::is_error in all error types"
  static_assert((... && is_error_v<AllowedErrors>),
  ^                     ~~~~~~~~~~~~~~~~~~~~~~~~~
../src/crimson/os/futurized_store.h:46:11: note: in instantiation of template class 'crimson::errorator<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >' requested here
  virtual read_errorator::future<ceph::bufferlist> read(
          ^
In file included from ../src/crimson/os/cyanstore/cyan_store.cc:4:
In file included from ../src/crimson/os/cyanstore/cyan_store.h:18:
In file included from ../src/crimson/os/futurized_store.h:14:
In file included from ../src/crimson/osd/exceptions.h:9:
../src/crimson/common/errorator.h:82:47: error: constexpr variable cannot have non-literal type 'const crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec>'
  static inline constexpr unthrowable_wrapper instance{};
                                              ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/9/../../../../include/c++/9/type_traits:1302:66: note: in instantiation of template class 'crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec>' requested here
    : public integral_constant<bool, __is_base_of(_Base, _Derived)>
                                                                 ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/9/../../../../include/c++/9/type_traits:2962:40: note: in instantiation of template class 'std::is_base_of<crimson::error_t<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >, crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >' requested here
  inline constexpr bool is_base_of_v = is_base_of<_Base, _Derived>::value;
                                       ^
../src/crimson/common/errorator.h:274:50: note: in instantiation of variable template specialization 'std::is_base_of_v<crimson::error_t<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >, crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >' requested here
  static inline constexpr bool is_error_v = std::is_base_of_v<error_t<T>, T>;
                                                 ^
../src/crimson/common/errorator.h:276:25: note: in instantiation of static data member 'crimson::errorator<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec>, crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >::is_error_v<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >' requested here
  static_assert((... && is_error_v<AllowedErrors>),
                        ^
../src/crimson/os/futurized_store.h:56:11: note: in instantiation of template class 'crimson::errorator<crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec>, crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec> >' requested here
  virtual get_attr_errorator::future<ceph::bufferptr> get_attr(
          ^
../src/crimson/common/errorator.h:82:47: note: incomplete type 'const crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec>' is not a literal type
  static inline constexpr unthrowable_wrapper instance{};
                                              ^
../src/crimson/common/errorator.h:80:8: note: definition of 'crimson::unthrowable_wrapper<const std::error_code &, &crimson::ec>' is not complete until the closing '}'
struct unthrowable_wrapper : error_t<unthrowable_wrapper<ErrorT, ErrorV>> {
       ^

...

Reply:
i see. a minimal producer could be found at https://www.godbolt.org/z/bcM3bW . in which, i suggest an alternative fix.

Reply:
Nice...
I'll fix the updated PR to include your change.